src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java

```diff
@@ -0,0 +1,2163 @@
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.stream.Collectors;
+
+import org.antlr.v4.runtime.BufferedTokenStream;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.grammar.java.CheckstyleJavaParser;
+import com.puppycrawl.tools.checkstyle.grammar.java.CheckstyleJavaParserBaseVisitor;
+import com.puppycrawl.tools.checkstyle.grammar.java.JavaLexer;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * Visitor class used to build Checkstyle's Java AST from the parse tree produced by
+ * {@link CheckstyleJavaParser}. In each {@code visit...} method, we visit the children of a node
+ * (which correspond to subrules) or create terminal nodes (tokens), and return a subtree as a
+ * result.
+ *
+ * <p>Example:</p>
+ *
+ * <p>The following package declaration:</p>
+ * <pre>
+ * package com.puppycrawl.tools.checkstyle;
+ * </pre>
+ *
+ * <p>
+ * Will be parsed by the {@code packageDeclaration} rule from {@code CheckstyleJavaParser.g4}:
+ * </p>
+ * <pre>
+ * packageDeclaration
+ *     : annotations[true] LITERAL_PACKAGE qualifiedName SEMI
+ *     ;
+ * </pre>
+ *
+ * <p>
+ * We override the {@code visitPackageDeclaration} method generated by ANTLR in
+ * {@code CheckstyleJavaParserBaseVisitor} at
+ * {@link JavaAstVisitor#visitPackageDeclaration(CheckstyleJavaParser.PackageDeclarationContext)}
+ * to create a subtree based on the subrules and tokens found in the {@code packageDeclaration}
+ * subrule accordingly, thus producing the following AST:
+ * </p>
+ * <pre>
+ * PACKAGE_DEF -&gt; package
+ * |--ANNOTATIONS -&gt; ANNOTATIONS
+ * |--DOT -&gt; .
+ * |   |--DOT -&gt; .
+ * |   |   |--DOT -&gt; .
+ * |   |   |   |--IDENT -&gt; com
+ * |   |   |   `--IDENT -&gt; puppycrawl
+ * |   |   `--IDENT -&gt; tools
+ * |   `--IDENT -&gt; checkstyle
+ * `--SEMI -&gt; ;
+ * </pre>
+ * <p>
+ * See https://github.com/checkstyle/checkstyle/pull/10434 for a good example of how
+ * to make changes to Checkstyle's grammar and AST.
+ * </p>
+ * <p>
+ * The order of {@code visit...} methods in {@code JavaAstVisitor.java} and production rules in
+ * {@code CheckstyleJavaParser.g4} should be consistent to ease maintenance.
+ * </p>
+ */
```

Comment:
Should I write a more detailed README/ guide with code examples on how to update grammar/ visitor? If so, should it live in the same directory as the parser and lexer grammar?